Feature - Composite Object AttributeSet Attribute

RossMurr4y opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
* Implement new properties `AttributeSet` and `Component` to the composite object structure.
  * `AttributeSet` defines a reference to an AttributeSet type.
  * `Component` defines a reference to another component. Intended for internal use only (the Engine will apply it where required).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
* Contributes to https://github.com/hamlet-io/engine/issues/1491
* Enable composite structure re-use
* Remove file dependency issues at file load time

**Example of Implementation:**
The attribute `Links` on this example component will at the time of loading require no other composite object structures exist in order to be defined. During engine runtime, was the `getCompositeObject()` function is called, it will search the structure of the `attributeset` schema for the `link` type and add its structure as children to the `Links` attribute. 

```ftl
[@addComponent
    type=EXAMPLE_COMPONENT_TYPE
    properties=[... ]
    attributes=
        [
            {
                "Names" : "Links",
                "Subobjects" : true,
                "AttributeSet" : LINK_ATTRIBUTESET_TYPE
            }
        ]
/]
```

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Local template generation.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions
<!---
    Are the changes mandatory (breaking) or optional?
    What changes must a consumer of this repository make in order to utilise it?
    Are there other issues or steps that need to happen once this PR is merged?

    Add a checklist of items or leave the default of "None"
-->
- [ ] Requires https://github.com/hamlet-io/engine/pull/1498

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] None of the above.

RossMurr4y commented:
Marked as draft until https://github.com/hamlet-io/engine/pull/1498 is merged, then needs a rebase before unmarking as draft.

roleyfoley commented:
Looks like a good refactor and should make understanding easier 

Only thing I would ask is do we need to support multiple schemas? I think the idea is that that AttributeSets are our answer to shared schema. Do we need to support sharing all schemas and how would that be implemented for something like bueprint?

roleyfoley commented on engine/base.ftl:
```diff
@@ -765,17 +773,74 @@ are added.
+                    [#case "attributeset" ]
+                        [#local attributeRefValue = attributeSetConfiguration[attribute.Reference.Type]!{} ]
+                        [#break]
+                    [#case "blueprint" ]
+                        [#local attributeRefValue = blueprintObject[attribute.Reference.Type]!{} ]
+                        [#break]
```
Not sure I can see when you would ref a blueprint, that would become a bit of inception. Since anything inside of this composite object is most likely going to be in the blueprint.

roleyfoley commented on engine/base.ftl:
```diff
@@ -765,17 +773,74 @@ are added.
+                    [#case "component" ]
+                        [#local attributeRefValue = componentConfiguration[attribute.Reference.Type]!{} ]
+                        [#break]
+                    [#case "reference" ]
+                        [#local attributeRefValue = referenceConfiguration[attribute.Reference.Type]!{} ]
+                        [#break]
```
Probably good to remove reference to, since components should reference an instance of reference data rather than include reference configuration in the component

RossMurr4y commented:
> Looks like a good refactor and should make understanding easier
> 
> Only thing I would ask is do we need to support multiple schemas? I think the idea is that that AttributeSets are our answer to shared schema. Do we need to support sharing all schemas and how would that be implemented for something like bueprint?
> 
> edit: After conversation I think we can still use an object, there is a strong requirement to show the relationships between components and subcomponents so this can stay as an object, but we should limit the available reference sets

After further conversation, @roleyfoley and I have determined that the "Reference" composite object attribute does not need to be an object at all. There is no need to reference types other than AttributeSet, and referencing child components is handled independent of the composite object structure.

I will update this PR to rename "Reference" attribute as "AttributeSet", which provides more clarity as to what is happening and why.

RossMurr4y commented:
@roleyfoley I have updated this draft and it is now ready for review.

As discussed, I have replaced the "Reference" composite object attribute with "AttributeSet" which is far simpler to assign and doesn't introduce unnecessary complexity.

At schema generation time, there is a need to be able to differentiate between "I am constructing a reference to an AttributeSet" and "I am constructing another kind of reference" (such as from one component to another for subComponents). For this I have added the "Component" composite object attribute, however this is not intended for use within definition files (id.ftl), it's used only by the engine itself as is recognises a subComponent and constructs a reference to another component.

roleyfoley commented:
@RossMurr4y I think this PR will need a rebase as it does include some changes made outside of this PR and is currently conflicting 

@ml019 Could you pass an eye over this one given the potential impact around getCompositeObject?

RossMurr4y commented:
Conflicts resolved and matching PR submitted for AWS provider. Azure does not require an update.

RossMurr4y commented:
Rebased and updated with changes from #1521

RossMurr4y commented:
@ml019 would you mind reviewing? Other repositories are beginning to have features merged that expect the changes implemented here.

ml019 commented on engine/base.ftl:
```diff
@@ -721,7 +721,8 @@ are added.
                     "SubObjects" : false,
-                    "PopulateMissingChildren" : true
+                    "PopulateMissingChildren" : true,
+                    "AttributeSet" : ""
```
Also needs an entry for Component?

RossMurr4y replied:
updated in https://github.com/hamlet-io/engine/pull/1502/commits/8e88a00f6a4cd65b69bf555e6c2325a03f45d7ed, cheers

ml019 commented on engine/base.ftl:
```diff
@@ -745,7 +746,9 @@ are added.
-                        "PopulateMissingChildren" : attribute.PopulateMissingChildren!true
+                        "PopulateMissingChildren" : attribute.PopulateMissingChildren!true,
+                        "AttributeSet" : attribute.AttributeSet!"",
+                        "Component" : attribute.Component!""
```
There is no mention of component in the PR description. I understand what an attributeSet is doing, but what is component doing?

RossMurr4y replied:
Component defines a SubComponent within the composite object model. It is not intended for further use, or manual assignment within `id.ftl` files, rather it separates a reference to another component schema out from a reference to an AttributeSet.

A method to differentiate the two is necessary to provide accurate schema references, that define both the "SchemaType" (component, reference, attributeset) and the Instance (baseline, alertprofile, link, etc) values.

RossMurr4y replied:
I've updated the PR description to include this.

ml019 commented on providers/shared/components/component.ftl:
```diff
@@ -131,165 +131,6 @@
 [#assign USERPOOL_RESOURCE_COMPONENT_TYPE = "userpoolresource" ]
 
-[#assign
-    filterChildrenConfiguration = [
```
A filter is basically what is now being referred to as a district. A link is basically a district with some extra attributes, which is why there were two separate definitions here.

It doesn't need addressing now but at some point, the link attribute set may need to be separated into two structure again.

RossMurr4y replied:
We can do that easily enough when the time comes, no?

ml019 commented:
Will need to be rebased as well
